src/python/pants/backend/python/lint/ruff/subsystem.py

```diff
@@ -30,17 +37,32 @@ class RuffMode(str, Enum):
 # --------------------------------------------------------------------------------------
 
 
-class Ruff(PythonToolBase):
+class Ruff(TemplatedExternalTool):
     options_scope = "ruff"
     name = "Ruff"
-    help_short = "The Ruff Python formatter (https://github.com/astral-sh/ruff)."
-
-    default_main = ConsoleScript("ruff")
-    default_requirements = ["ruff>=0.1.2,<1"]
-
-    register_interpreter_constraints = True
-
-    default_lockfile_resource = ("pants.backend.python.lint.ruff", "ruff.lock")
+    help = "The Ruff Python formatter (https://github.com/astral-sh/ruff)."
+
+    default_version = "0.4.9"
+    default_known_versions = [
+        # Custom URls because the tagging scheme changed from `v0.4.x` (Note: v) to `0.5.x`:
+        "0.4.9|macos_arm64|5f4506d7ec2ae6ac5a48ba309218a4b825a00d4cad9967b7bbcec1724ef04930|8148128|https://github.com/astral-sh/ruff/releases/download/v0.4.9/ruff-0.4.9-aarch64-apple-darwin.tar.gz",
+        "0.4.9|macos_x86_64|e4d745adb0f5a0b08f2c9ca71e57f451a9b8485ae35b5555d9f5d20fc93a6cb6|8510706|https://github.com/astral-sh/ruff/releases/download/v0.4.9/ruff-0.4.9-x86_64-apple-darwin.tar.gz",
+        "0.4.9|linux_arm64|00c50563f9921a141ddd4ec0371149f3bbfa0369d9d238a143bcc3a932363785|8106747|https://github.com/astral-sh/ruff/releases/download/v0.4.9/ruff-0.4.9-aarch64-unknown-linux-musl.tar.gz",
+        "0.4.9|linux_x86_64|5ceba21dad91e3fa05056ca62f278b0178516cfad8dbf08cf2433c6f1eeb92d3|8863118|https://github.com/astral-sh/ruff/releases/download/v0.4.9/ruff-0.4.9-x86_64-unknown-linux-musl.tar.gz",
```

Comment:
So in future, upgrading Ruff means updating these definitions right?

Reply:
Yeah, following https://www.pantsbuild.org/2.23/docs/contributions/development/maintenance-tasks-and-scripts#external-tools-downloaded-executables, for instance.

I'd be imagining we _append_ new versions here, rather than replace, so that someone who's pinned to (for example) 0.4.9 in Pants 2.23 via `version = "0.4.9"` can upgrade to Pants 2.24, 2.25, ... without having to change their Ruff config at all.

For instance, how terraform's maintained: https://github.com/pantsbuild/pants/pull/20469 (although maybe including _all_ patch releases is a bit unnecessary for ruff).

Reply:
Yes, we append!